# Laboratorio 5
Universidad del Valle de Guatemala <br>
Security Data Science <br>
Pablo Andrés Zamora Vásquez <br>
Carné 21780

## Parte 1 - Filtrado y preprocesamiento

In [1]:
import json
import random
import pandas as pd
from tqdm import tqdm

*1. Cargue la información del archivo large_even.json en una lista, muestre la cantidad de
registros total (deben ser 746, 909). Este es nuestro tráfico inicial*

In [2]:
# Cargar los datos
with open('large_eve.json') as f:
    data = [json.loads(line) for line in f]

# Mostrar la cantidad total de registros
print(f'Total de registros: {len(data)}')  # Debería ser 746909

Total de registros: 746909


*2. Cargue únicamente aquellos registros que sean DNS.*

*3. Muestre la nueva cantidad de registros filtrados.*

In [3]:
# Filtrar solo los eventos de tipo 'dns'
dns_records = [record for record in data if record.get('dns')]

# Mostrar cantidad de registros DNS
print(f'Registros DNS: {len(dns_records)}')  # Deberían ser 21484


Registros DNS: 15749


*4. Muestre la información de 2 registros cualesquiera*

In [4]:
# Mostrar dos registros aleatorios de los DNS
for r in random.sample(dns_records, 2):
    print(json.dumps(r, indent=2))

{
  "timestamp": "2017-07-22T19:05:40.387366-0500",
  "flow_id": 548497561807142,
  "pcap_cnt": 2364628,
  "event_type": "dns",
  "vlan": 150,
  "src_ip": "192.168.205.170",
  "src_port": 41924,
  "dest_ip": "192.168.207.4",
  "dest_port": 53,
  "proto": "UDP",
  "dns": {
    "type": "query",
    "id": 56010,
    "rrname": "AOLDTCMA04.office.aol.com",
    "rrtype": "A",
    "tx_id": 0
  }
}
{
  "timestamp": "2017-07-22T18:29:17.388339-0500",
  "flow_id": 758961553665018,
  "pcap_cnt": 504614,
  "event_type": "dns",
  "vlan": 150,
  "src_ip": "192.168.207.4",
  "src_port": 53,
  "dest_ip": "192.168.205.188",
  "dest_port": 53277,
  "proto": "UDP",
  "dns": {
    "type": "answer",
    "id": 45092,
    "rcode": "NXDOMAIN",
    "rrname": "mirror.its.uidaho.edu"
  }
}


*5. Debido a que la data consiste en estructuras JSON anidadas, utilice la característica
json_normalize para normalizar la información y asignarla en un dataframe.*

In [4]:
from pandas import json_normalize

df_dns = json_normalize(dns_records)
print(f'Shape del DataFrame: {df_dns.shape}')  # Debería ser (21484, 163)


Shape del DataFrame: (15749, 18)


*6. Como estamos buscando dominios DGA, debemos filtrar los registros DNS para aquellos
registros tipo A.*

In [5]:
# Filtrar por registros tipo A
df_dns_type_a = df_dns[df_dns['dns.rrtype'] == 'A']
print(f'Registros DNS tipo A: {len(df_dns_type_a)}')  # Deberían ser 2849


Registros DNS tipo A: 2849


*7. Filtre los dominios únicos*

In [6]:
# Filtrar los dominios únicos
unique_domains = df_dns_type_a['dns.rrname'].dropna().unique()
print(f'Dominios únicos tipo A: {len(unique_domains)}')  # Deberían ser 177

Dominios únicos tipo A: 177


*8. Escriba una función que obtenga el TLD para un dominio.*

**Prompt realizado a GPT-4o:** Escribe una función en Python que reciba un dominio completo como 'api.wunderground.com' y devuelva el top-level domain (TLD), en este caso 'wunderground.com'.

In [7]:
def extract_tld(domain):
    parts = domain.strip('.').split('.')
    if len(parts) >= 2:
        return '.'.join(parts[-2:])
    else:
        return domain

*9. Del dataframe de dominios únicos de tipo A, obtenga el TLD (top level domain) utilizando la
función anterior para crear una columna nueva llamada domain_tld, y elimine todas las demás
columnas.*

In [8]:
df_unique_domains = pd.DataFrame(unique_domains, columns=['domain'])
df_unique_domains['domain_tld'] = df_unique_domains['domain'].apply(extract_tld)

print(df_unique_domains.head())

                                       domain        domain_tld
0                        api.wunderground.com  wunderground.com
1                         stork79.dropbox.com       dropbox.com
2  hpca-tier2.office.aol.com.ad.aol.aoltw.net         aoltw.net
3        safebrowsing.clients.google.com.home          com.home
4                         fxfeeds.mozilla.com       mozilla.com


## Parte 2 - Data Science

*10. Utilice Gemini para clasificar los dominios como DGA (1) o legítimos (0)*

In [56]:
import google.generativeai as genai

In [ ]:
genai.configure(api_key='AIzaSyCjQXdW7j2L9XVt85RdV1k1XWWNkjLAq48')

# Configuración para que no bloquee prompts con términos como "DGA", "malware", "phishing", etc.

safe = [
        {
            "category": "HARM_CATEGORY_DANGEROUS",
            "threshold": "BLOCK_NONE",
        },
        {
            "category": "HARM_CATEGORY_HARASSMENT",
            "threshold": "BLOCK_NONE",
        },
        {
            "category": "HARM_CATEGORY_HATE_SPEECH",
            "threshold": "BLOCK_NONE",
        },
        {
            "category": "HARM_CATEGORY_SEXUALLY_EXPLICIT",
            "threshold": "BLOCK_NONE",
        },
        {
            "category": "HARM_CATEGORY_DANGEROUS_CONTENT",
            "threshold": "BLOCK_NONE",
        },
    ]

In [61]:
model = genai.GenerativeModel('gemini-1.5-flash-001')

zero_shot_prompt = f"""
You are a security analyst and you have been tasked with analyzing a list of domains to determine if they are malicious or not.
Classify each domain as:
- 1: If you think it's generated by an algorithm (DGA)
- 0: If you think it's a legitimate domain, like google.com, amazonaws.com, etc.

Give your response as a list of tuples separated by a coma, where each tuple contains the domain and its classification (domain, 1/0).
Don't include anything else in your response.
Domains:
"""

In [69]:
import ast

# Enviar en bloques de 20
def classify_with_gemini(domains):
    results = []
    for i in range(0, len(domains), 20):
        batch = domains[i:i+20]
        input_text = zero_shot_prompt + "\n".join(batch)
        response = model.generate_content(input_text, safety_settings=safe)
        # Procesar el output, por ejemplo: [('xkfjds.com', 1), ('google.com', 0)]
        try:
            fixed_text = "[" + response.text.strip().rstrip(',') + "]"
            output = ast.literal_eval(fixed_text)
            results.extend(output)
        except:
            print("Error procesando respuesta:", response.text)
    return results

In [71]:
tlds = df_unique_domains['domain_tld'].tolist()
classified = classify_with_gemini(tlds)

# Convertir a DataFrame
df_classified = pd.DataFrame(classified, columns=['domain_tld', 'gemini_label'])

# Unirlo con el dataframe original
df_domains_final = df_unique_domains.merge(df_classified, on='domain_tld', how='left')


In [72]:
print(df_domains_final.head())

                                       domain        domain_tld  gemini_label
0                        api.wunderground.com  wunderground.com           0.0
1                         stork79.dropbox.com       dropbox.com           0.0
2  hpca-tier2.office.aol.com.ad.aol.aoltw.net         aoltw.net           1.0
3  hpca-tier2.office.aol.com.ad.aol.aoltw.net         aoltw.net           1.0
4  hpca-tier2.office.aol.com.ad.aol.aoltw.net         aoltw.net           1.0


*11. Filtre aquellos considerados como DGA (valor 1 ) y muéstrelos. Después de eliminar
duplicados, debe obtener 61 registros únicos.*

In [ ]:
# Contar los dominios clasificados como DGA
dga_domains = df_classified[df_classified['gemini_label'] == 1]
print(f'Dominios clasificados como DGA: {len(dga_domains)}')

Dominios clasificados como DGA: 41


Aunque se obtuvieron menos registros de los esperados, la clasificación realizada por el LLM parece tener sentido, ya que TLDs como *dropbox.com* o *mozilla.com* **fueron clasificados como legítimos**, mientras que dominios como *vtlfccmfxlkgifuf.com* **fueron clasificados como sospechosos**.

In [ ]:
# Guardar los dominios clasificados por Gemini en un archivo CSV
df_classified.to_csv('domains_classification.csv', index=False)

## Parte 3

*12. Escriba una función que utilice la lista de un millón de TLD proporcionada en Canvas, y
devuelva 0 si el TLD se encuentra en la lista y 1 si no está.*

In [ ]:
# Cargar el archivo CSV de dominios clasificados por Gemini
df_classified = pd.read_csv('domains_classification.csv')

In [10]:
# Cargar el top-1m.csv
top1m_df = pd.read_csv('top-1m.csv', header=None, names=['rank', 'domain'])

# Convertir a set para búsqueda rápida
top1m_set = set(top1m_df['domain'])

**Prompt realizado a GPT-4o:** Escribe una función en Python que reciba un dominio (TLD) y verifica si está en una lista cargada previamente (top-1m.csv). La función debe devolver 0 si está en la lista (legítimo), y 1 si no está. La lista no debe recargarse cada vez.

In [11]:
def is_not_in_top1m(domain_tld, top1m_domains=top1m_set):
    return 0 if domain_tld in top1m_domains else 1

*13. Utilice la función para determinar si los TLD se encuentran en dicha lista. Filtre aquellos que si
se encuentran*

In [16]:
dga_domains = df_classified[df_classified['gemini_label'] == 1].copy()

# Aplicar la función
dga_domains['not_in_top1m'] = dga_domains['domain_tld'].apply(is_not_in_top1m)

# Filtrar los que NO están en top-1m
dga_suspects = dga_domains[dga_domains['not_in_top1m'] == 1]

# Eliminar duplicados si aún hay
dga_suspects = dga_suspects.drop_duplicates(subset='domain_tld')

print(f'Dominios sospechosos finales: {len(dga_suspects)}')  # Deberían ser 13
dga_suspects[['domain_tld']]

Dominios sospechosos finales: 27


domain_tld
2                    aoltw.net
3                     com.home
13                   wpad.home
14              stayonline.net
18            22.110phpmyadmin
19            informaction.com
20             com.localdomain
24   110phpmyadmin.localdomain
27               ntkrnlpa.info
39                         cn.
63              hackerlabs.vpn
64           bigflickrfeed.com
66                        wpad
67                     com.lan
71                     206.56"
88                     26-27.0
126                         FL
131                    saruman
135                    21.1201
150      whitecell.localdomain
159       vtlfccmfxlkgifuf.com
162                     22.254
164                   254.home
165                    21-28.0
166       ejfodfmfxlkgifuf.xyz
167                     0.home
168                    22.201:

*14. Cree una función que, en base al TLD, devuelva la fecha de creación de este.*

**Prompt realizado a GPT-4o:** Escribe una función en Python que reciba un dominio y devuelva la fecha de creación utilizando la librería whois. Si hay errores o no hay información, debe retornar 'Error' o 'No info'.

In [17]:
import whois
from datetime import datetime

def get_creation_date(domain):
    try:
        info = whois.whois(domain)
        creation_date = info.creation_date

        # Puede ser una lista o un solo valor
        if isinstance(creation_date, list):
            creation_date = creation_date[0]

        return creation_date.strftime('%Y-%m-%d') if creation_date else 'No info'
    
    except Exception as e:
        return 'Error'

In [ ]:
# Aplicar con barra de progreso
dga_suspects['creation_date'] = [
    get_creation_date(domain) for domain in tqdm(dga_suspects['domain_tld'])
]

100%|██████████| 27/27 [00:06<00:00,  4.02it/s]


*15. Muestre la fecha de creación para cada uno de los 13 dominios finales ¿Cuáles son los
dominios que podemos confirmar como sospechosos?*

In [19]:
dga_suspects[['domain_tld', 'creation_date']]

domain_tld creation_date
2                    aoltw.net    2000-01-10
3                     com.home       No info
13                   wpad.home       No info
14              stayonline.net    1999-12-14
18            22.110phpmyadmin       No info
19            informaction.com    1998-11-25
20             com.localdomain       No info
24   110phpmyadmin.localdomain       No info
27               ntkrnlpa.info       No info
39                         cn.       No info
63              hackerlabs.vpn       No info
64           bigflickrfeed.com         Error
66                        wpad       No info
67                     com.lan       No info
71                     206.56"       No info
88                     26-27.0       No info
126                         FL       No info
131                    saruman       No info
135                    21.1201       No info
150      whitecell.localdomain       No info
159       vtlfccmfxlkgifuf.com         Error
162                     22.254       No info
164                   254.home       No info
165                    21-28.0       No info
166       ejfodfmfxlkgifuf.xyz       No info
167                     0.home       No info
168                    22.201:       No info

Pueden considerarse como confiables aquellos dominios que no hayan sido creados recientemente, puesto que esto indica que es menos probable que hayan sido generados por un DGA. Los siguientes dominios parecen ser confiables, ya que tienen una fecha de creación antigua:

- aoltw.net → 2000-01-10
- stayonline.net → 1999-12-14
- informaction.com → 1998-11-25

Por otro lado, el resto de dominios no devuelve información de su fecha de creación (*No info*) o simplemente devuelven error, por lo que pueden considerarse como sospechosos, ya que esto indica que el dominio no está registrado o no es válido.
<hr>

*16. Recuerde que los dominios DGA son conocidos por formarse de forma aleatoria: secuencias
aleatorias de caracteres, no palabras. Indique que dominios sospechosos tienen este patrón y
que pueden confirmarse como dominios DGA*

Dado que los dominios DGA pueden distinguirse gracias a que se generan de manera aleatoria, es posible analizar los dominios anteriores en busca de patrones sospechosos. Entre ellos, los más evidentes son:

- vtlfccmfxlkgifuf.com
- ejfodfmfxlkgifuf.xyz

Tomando en cuenta la información de fecha de la creación de los dominios y la búsqueda de patrones sospechosos, también es posible clasificar los siguientes dominios como sospechosos:

- 206.56", 21.1201, 22.201:, 26-27.0 → Estructuras de tipo IP
- com.home, 0.home, 254.home → Sin información de fecha de creación y uso de "home" (no usado en dominios públicos)
- 110phpmyadmin.localdomain → Sin información de fecha de creación y uso de palabras técnicas (phpmyadmin, localdomain)
- ntkrnlpa.info → Patrón sospechoso, además se asemeja a "kernel" (un malware típico)
- wpad.home → Sin información de fecha de creación, uso de "home" y de "wpad" (protocolo que permite encontrar un archivo de configuración proxy)